File: packages/core-ts/src/limits.ts
/**
 * Normative caps, loaded from the shared contract rather than hardcoded.
 *
 * Both language cores read `contracts/v1/limits.json`. A deployment may lower a cap
 * (`narrowLimits`); raising one is rejected here so a config file can never widen the
 * security envelope that the acceptance gates measure.
 */
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const here = dirname(fileURLToPath(import.meta.url));

/** Works from `src/` (checkout) and from `dist/` (installed package). */
export function contractsDir(): string {
  return join(here, "..", "contracts", "v1");
}

/** The normative store DDL, read from the vendored contract rather than embedded. */
export function storeDdlPath(): string {
  return join(here, "..", "contracts", "store", "v1.sql");
}

export function storeDdl(): string {
  return readFileSync(storeDdlPath(), "utf8");
}

function loadJson<T>(name: string): T {
  return JSON.parse(readFileSync(join(contractsDir(), name), "utf8")) as T;
}

export interface RawLimits {
  reader_model: string;
  contract: { emitted_version: string; supported_request_versions: string[] };
  gate: Record<string, number>;
  bytes: Record<string, number>;
  counts: Record<string, number>;
  tokens: Record<string, number>;
  deadlines_ms: Record<string, number>;
  spill: { ttl_seconds: number; dir_mode: string; file_mode: string };
  json: Record<string, number>;
  inspect: Record<string, number>;
  disclosure: Record<string, number>;
  store: Record<string, number | string>;
  accounting: Record<string, number | string>;
}

export interface StatusCodePairs {
  statuses: string[];
  codes: string[];
  pairs: Record<string, string[]>;
}

let rawCache: RawLimits | undefined;
let pairsCache: StatusCodePairs | undefined;

export function rawLimits(): RawLimits {
  rawCache ??= loadJson<RawLimits>("limits.json");
  return rawCache;
}

export function statusCodePairs(): StatusCodePairs {
  pairsCache ??= loadJson<StatusCodePairs>("status-code-pairs.json");
  return pairsCache;
}

export interface Limits {
  readonly readerModel: string;
  readonly fullReadMaxLines: number;
  readonly targetedReadMaxLines: number;
  readonly targetedSearchMaxMatches: number;
  readonly probeMaxLinesScanned: number;
  readonly maxToolResultBytes: number;
  readonly maxEnvelopeBytes: number;
  readonly maxTargetedReadBytes: number;
  readonly maxSourceBytes: number;
  readonly maxChunkBytes: number;
  readonly maxAnswerBytes: number;
  readonly maxQuoteBytes: number;
  readonly maxQuestionBytes: number;
  readonly sessionSpillQuotaBytes: number;
  readonly maxClaimTextBytes: number;
  readonly maxSourcesPerRequest: number;
  readonly maxChunksPerRequest: number;
  readonly maxCitations: number;
  readonly maxConcurrentModelCalls: number;
  readonly maxChunkOverlapLines: number;
  readonly maxTransientRetries: number;
  readonly maxClaimsPerAnswer: number;
  readonly maxCitationIdsPerClaim: number;
  readonly maxFormatRetries: number;
  readonly maxChunkTokens: number;
  readonly maxRequestInputTokens: number;
  readonly maxOutputTokensPerCall: number;
  readonly bytesPerTokenEstimate: number;
  readonly gateProbeDeadlineMs: number;
  readonly spillIoDeadlineMs: number;
  readonly modelCallDeadlineMs: number;
  readonly requestDeadlineMs: number;
  readonly spillTtlSeconds: number;
  readonly jsonMaxDepth: number;
  readonly jsonMaxNodes: number;
  // -- 1.1 additions --
  readonly maxExtendedEnvelopeBytes: number;
  readonly maxExtractionBytes: number;
  readonly inspectMaxResultBytes: number;
  readonly inspectMaxSegments: number;
  readonly inspectMaxLinesPerPage: number;
  readonly inspectMaxBytesPerPage: number;
  readonly inspectMaxScanLines: number;
  readonly inspectMaxScanBytes: number;
  readonly inspectMaxSearchMatches: number;
  readonly inspectMaxNeedleBytes: number;
  readonly disclosureMaxPerSourceBytes: number;
  readonly disclosureMaxPerSessionBytes: number;
  readonly storeDdlVersion: number;
  readonly storeBusyTimeoutMs: number;
  readonly storeMaxEntries: number;
  readonly storeMaxBytes: number;
  readonly storeHandleTtlSeconds: number;
  readonly statsMaxRecordsPerPage: number;
  readonly statsMaxPages: number;
}

function pick(group: Record<string, number>, key: string): number {
  const value = group[key];
  if (typeof value !== "number") {
    throw new Error(`limits.json is missing ${key}`);
  }
  return value;
}

export function defaultLimits(): Limits {
  const raw = rawLimits();
  return Object.freeze({
    readerModel: raw.reader_model,
    fullReadMaxLines: pick(raw.gate, "full_read_max_lines"),
    targetedReadMaxLines: pick(raw.gate, "targeted_read_max_lines"),
    targetedSearchMaxMatches: pick(raw.gate, "targeted_search_max_matches"),
    probeMaxLinesScanned: pick(raw.gate, "probe_max_lines_scanned"),
    maxToolResultBytes: pick(raw.bytes, "max_tool_result_bytes"),
    maxEnvelopeBytes: pick(raw.bytes, "max_envelope_bytes"),
    maxTargetedReadBytes: pick(raw.bytes, "max_targeted_read_bytes"),
    maxSourceBytes: pick(raw.bytes, "max_source_bytes"),
    maxChunkBytes: pick(raw.bytes, "max_chunk_bytes"),
    maxAnswerBytes: pick(raw.bytes, "max_answer_bytes"),
    maxQuoteBytes: pick(raw.bytes, "max_quote_bytes"),
    maxQuestionBytes: pick(raw.bytes, "max_question_bytes"),
    sessionSpillQuotaBytes: pick(raw.bytes, "session_spill_quota_bytes"),
    maxClaimTextBytes: pick(raw.bytes, "max_claim_text_bytes"),
    maxSourcesPerRequest: pick(raw.counts, "max_sources_per_request"),
    maxChunksPerRequest: pick(raw.counts, "max_chunks_per_request"),
    maxCitations: pick(raw.counts, "max_citations"),
    maxConcurrentModelCalls: pick(raw.counts, "max_concurrent_model_calls"),
    maxChunkOverlapLines: pick(raw.counts, "max_chunk_overlap_lines"),
    maxTransientRetries: pick(raw.counts, "max_transient_retries"),
    maxClaimsPerAnswer: pick(raw.counts, "max_claims_per_answer"),
    maxCitationIdsPerClaim: pick(raw.counts, "max_citation_ids_per_claim"),
    maxFormatRetries: pick(raw.counts, "max_format_retries"),
    maxChunkTokens: pick(raw.tokens, "max_chunk_tokens"),
    maxRequestInputTokens: pick(raw.tokens, "max_request_input_tokens"),
    maxOutputTokensPerCall: pick(raw.tokens, "max_output_tokens_per_call"),
    bytesPerTokenEstimate: pick(raw.tokens, "bytes_per_token_estimate"),
    gateProbeDeadlineMs: pick(raw.deadlines_ms, "gate_probe"),
    spillIoDeadlineMs: pick(raw.deadlines_ms, "spill_io"),
    modelCallDeadlineMs: pick(raw.deadlines_ms, "model_call"),
    requestDeadlineMs: pick(raw.deadlines_ms, "request"),
    spillTtlSeconds: raw.spill.ttl_seconds,
    jsonMaxDepth: pick(raw.json, "max_depth"),
    jsonMaxNodes: pick(raw.json, "max_nodes"),
    maxExtendedEnvelopeBytes: pick(raw.bytes, "max_extended_envelope_bytes"),
    maxExtractionBytes: pick(raw.bytes, "max_extraction_bytes"),
    inspectMaxResultBytes: pick(raw.inspect, "max_result_bytes"),
    inspectMaxSegments: pick(raw.inspect, "max_segments"),
    inspectMaxLinesPerPage: pick(raw.inspect, "max_lines_per_page"),
    inspectMaxBytesPerPage: pick(raw.inspect, "max_bytes_per_page"),
    inspectMaxScanLines: pick(raw.inspect, "max_scan_lines"),
    inspectMaxScanBytes: pick(raw.inspect, "max_scan_bytes"),
    inspectMaxSearchMatches: pick(raw.inspect, "max_search_matches"),
    inspectMaxNeedleBytes: pick(raw.inspect, "max_needle_bytes"),
    disclosureMaxPerSourceBytes: pick(raw.disclosure, "max_per_source_bytes"),
    disclosureMaxPerSessionBytes: pick(raw.disclosure, "max_per_session_bytes"),
    storeDdlVersion: pick(raw.store as Record<string, number>, "ddl_version"),
    storeBusyTimeoutMs: pick(raw.store as Record<string, number>, "busy_timeout_ms"),
    storeMaxEntries: pick(raw.store as Record<string, number>, "max_entries"),
    storeMaxBytes: pick(raw.store as Record<string, number>, "max_bytes"),
    storeHandleTtlSeconds: pick(raw.store as Record<string, number>, "handle_ttl_seconds"),
    statsMaxRecordsPerPage: pick(
      raw.accounting as Record<string, number>,
      "max_stats_records_per_page",
    ),
    statsMaxPages: pick(raw.accounting as Record<string, number>, "max_stats_pages"),
  });
}

export const DEFAULT_LIMITS: Limits = defaultLimits();
export const READER_MODEL = DEFAULT_LIMITS.readerModel;

/**
 * Contract revision. Two constants, deliberately separate: `EMITTED_SCHEMA_VERSION` is
 * what every envelope this core builds declares, `SUPPORTED_REQUEST_VERSIONS` is what it
 * will accept on input. Revision 1.1 is backward compatible - a 1.0 request is still
 * accepted and a 1.0 envelope still validates - but a request that declares 1.0 while
 * carrying a 1.1 field is refused rather than accepted with the field ignored.
 */
export const EMITTED_SCHEMA_VERSION: string = rawLimits().contract.emitted_version;
export const SUPPORTED_REQUEST_VERSIONS: ReadonlySet<string> = new Set(
  rawLimits().contract.supported_request_versions,
);
/** Backward-compatible alias. New code should say which of the two it means. */
export const SCHEMA_VERSION: string = EMITTED_SCHEMA_VERSION;

/** Fields and operations that only exist from 1.1 onward. */
export const V11_ONLY_REQUEST_FIELDS: ReadonlySet<string> = new Set(["refined"]);
export const V11_ONLY_OPERATIONS: ReadonlySet<string> = new Set(["inspect", "stats"]);
export const V11_ONLY_ENVELOPE_FIELDS: ReadonlySet<string> = new Set([
  "result_kind", "provenance", "accounting_id", "extraction", "stats", "recovery",
  "import_receipt",
]);

/** Deterministic estimator name recorded whenever provider usage is unavailable. */
export const BASELINE_ESTIMATE_METHOD = String(rawLimits().accounting["baseline_method"]);

export function supportedRequestVersion(version: unknown): boolean {
  return typeof version === "string" && SUPPORTED_REQUEST_VERSIONS.has(version);
}

/**
 * The serialized cap that applies to one envelope.
 *
 * Deterministic extraction and stats carry a bounded payload of their own - up to
 * `maxExtractionBytes` of exact snapshot bytes, or one page of operation records - so they
 * are measured against `maxExtendedEnvelopeBytes`. Every other envelope keeps the original
 * 16 KiB cap. Both values live in `contracts/v1/limits.json`.
 */
export function envelopeByteCap(
  resultKind: string | undefined,
  limits: Limits = DEFAULT_LIMITS,
): number {
  return resultKind === "deterministic_extraction" || resultKind === "stats"
    ? limits.maxExtendedEnvelopeBytes
    : limits.maxEnvelopeBytes;
}

/** Caps a deployment may not set to zero, because zero would disable rather than tighten. */
export const POSITIVE_LIMITS: ReadonlySet<string> = new Set([
  "bytesPerTokenEstimate", "maxChunkBytes", "maxChunkTokens", "maxConcurrentModelCalls",
  "inspectMaxResultBytes", "inspectMaxScanLines", "storeDdlVersion", "storeBusyTimeoutMs",
  "statsMaxRecordsPerPage", "statsMaxPages",
]);

/** Chunk byte budget: the smaller of the byte cap and the token cap in bytes. */
export function chunkByteBudget(limits: Limits = DEFAULT_LIMITS): number {
  return Math.min(limits.maxChunkBytes, limits.maxChunkTokens * limits.bytesPerTokenEstimate);
}

export function narrowLimits(limits: Limits, overrides: Partial<Record<keyof Limits, number>>): Limits {
  const next: Record<string, unknown> = { ...limits };
  for (const [key, value] of Object.entries(overrides)) {
    if (!(key in limits)) throw new Error(`unknown limit: ${key}`);
    const current = (limits as unknown as Record<string, unknown>)[key];
    if (typeof current !== "number" || typeof value !== "number") {
      throw new Error(`limit is not numeric: ${key}`);
    }
    if (!Number.isSafeInteger(value)) throw new Error(`limit is not a safe integer: ${key}`);
    if (value > current) throw new Error(`limit ${key} may only be narrowed (max ${current})`);
    if (value < 0) throw new Error(`limit ${key} may not be negative`);
    if (value === 0 && POSITIVE_LIMITS.has(key)) {
      throw new Error(`limit ${key} must be positive`);
    }
    next[key] = value;
  }
  return Object.freeze(next) as unknown as Limits;
}

export function legalPair(status: string, code: string): boolean {
  const pairs = statusCodePairs().pairs;
  const allowed = pairs[status];
  return Array.isArray(allowed) && allowed.includes(code);
}

File: packages/core-py/src/context_shunt/limits.py
"""Normative caps, loaded from the shared contract rather than hardcoded.

Both language cores read ``contracts/v1/limits.json``. A deployment may lower a cap
(``Limits.narrow``); raising one is rejected here so a config file can never widen the
security envelope that the acceptance gates measure.

Contract revision
-----------------
Two version constants, deliberately separate:

* :data:`EMITTED_SCHEMA_VERSION` - the revision every envelope this core builds declares.
* :data:`SUPPORTED_REQUEST_VERSIONS` - the revisions this core will accept on input.

Revision 1.1 is backward compatible: a 1.0 request is still accepted unchanged, and a 1.0
envelope still validates. What 1.1 adds is mandatory *for envelopes that declare 1.1* -
``result_kind``, ``provenance`` and ``accounting_id`` - plus the optional extraction,
stats and recovery blocks and the inspect/stats operations. A request that declares 1.0
and carries a 1.1 field is rejected rather than accepted with the field ignored.

:data:`SCHEMA_VERSION` is kept as an alias of the emitted revision so existing call sites
keep working; new code should say which of the two it means.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from functools import cache
from pathlib import Path
from typing import Any

CONTRACTS_ROOT = Path(__file__).resolve().parent / "contracts"
CONTRACTS_DIR = CONTRACTS_ROOT / "v1"
STORE_DDL_PATH = CONTRACTS_ROOT / "store" / "v1.sql"


@cache
def _load(name: str) -> dict[str, Any]:
    with (CONTRACTS_DIR / name).open("rb") as fh:
        return json.load(fh)


@cache
def raw_limits() -> dict[str, Any]:
    return _load("limits.json")


@cache
def status_code_pairs() -> dict[str, Any]:
    return _load("status-code-pairs.json")


EMITTED_SCHEMA_VERSION: str = raw_limits()["contract"]["emitted_version"]
SUPPORTED_REQUEST_VERSIONS: frozenset[str] = frozenset(
    raw_limits()["contract"]["supported_request_versions"]
)
#: Backward-compatible alias. Prefer the explicit constant that says which side you mean.
SCHEMA_VERSION: str = EMITTED_SCHEMA_VERSION

#: Fields that only exist from 1.1 onward. A request declaring an earlier revision that
#: carries one of these is refused, so an unknown mandatory field is never ignored.
V11_ONLY_REQUEST_FIELDS: frozenset[str] = frozenset({"refined"})
V11_ONLY_OPERATIONS: frozenset[str] = frozenset({"inspect", "stats"})
V11_ONLY_ENVELOPE_FIELDS: frozenset[str] = frozenset(
    {
        "result_kind",
        "provenance",
        "accounting_id",
        "extraction",
        "stats",
        "recovery",
        "import_receipt",
    }
)


def store_ddl() -> str:
    """The normative store DDL, read from the vendored contract."""
    return STORE_DDL_PATH.read_text(encoding="utf-8")


@dataclass(frozen=True)
class Limits:
    """Effective caps for one deployment. Frozen: nothing mutates caps at runtime."""

    reader_model: str
    full_read_max_lines: int
    targeted_read_max_lines: int
    targeted_search_max_matches: int
    probe_max_lines_scanned: int
    max_tool_result_bytes: int
    max_envelope_bytes: int
    max_extended_envelope_bytes: int
    max_targeted_read_bytes: int
    max_extraction_bytes: int
    max_source_bytes: int
    max_chunk_bytes: int
    max_answer_bytes: int
    max_quote_bytes: int
    max_question_bytes: int
    session_spill_quota_bytes: int
    max_claim_text_bytes: int
    max_sources_per_request: int
    max_chunks_per_request: int
    max_citations: int
    max_concurrent_model_calls: int
    max_chunk_overlap_lines: int
    max_transient_retries: int
    max_claims_per_answer: int
    max_citation_ids_per_claim: int
    max_format_retries: int
    max_chunk_tokens: int
    max_request_input_tokens: int
    max_output_tokens_per_call: int
    bytes_per_token_estimate: int
    gate_probe_deadline_ms: int
    spill_io_deadline_ms: int
    model_call_deadline_ms: int
    request_deadline_ms: int
    spill_ttl_seconds: int
    json_max_depth: int
    json_max_nodes: int
    # -- 1.1 additions ---------------------------------------------------
    inspect_max_result_bytes: int
    inspect_max_segments: int
    inspect_max_lines_per_page: int
    inspect_max_bytes_per_page: int
    inspect_max_scan_lines: int
    inspect_max_scan_bytes: int
    inspect_max_search_matches: int
    inspect_max_needle_bytes: int
    disclosure_max_per_source_bytes: int
    disclosure_max_per_session_bytes: int
    store_ddl_version: int
    store_busy_timeout_ms: int
    store_max_entries: int
    store_max_bytes: int
    store_handle_ttl_seconds: int
    stats_max_records_per_page: int
    stats_max_pages: int

    @classmethod
    def defaults(cls) -> Limits:
        raw = raw_limits()
        return cls(
            reader_model=raw["reader_model"],
            full_read_max_lines=raw["gate"]["full_read_max_lines"],
            targeted_read_max_lines=raw["gate"]["targeted_read_max_lines"],
            targeted_search_max_matches=raw["gate"]["targeted_search_max_matches"],
            probe_max_lines_scanned=raw["gate"]["probe_max_lines_scanned"],
            max_tool_result_bytes=raw["bytes"]["max_tool_result_bytes"],
            max_envelope_bytes=raw["bytes"]["max_envelope_bytes"],
            max_extended_envelope_bytes=raw["bytes"]["max_extended_envelope_bytes"],
            max_targeted_read_bytes=raw["bytes"]["max_targeted_read_bytes"],
            max_extraction_bytes=raw["bytes"]["max_extraction_bytes"],
            max_source_bytes=raw["bytes"]["max_source_bytes"],
            max_chunk_bytes=raw["bytes"]["max_chunk_bytes"],
            max_answer_bytes=raw["bytes"]["max_answer_bytes"],
            max_quote_bytes=raw["bytes"]["max_quote_bytes"],
            max_question_bytes=raw["bytes"]["max_question_bytes"],
            session_spill_quota_bytes=raw["bytes"]["session_spill_quota_bytes"],
            max_claim_text_bytes=raw["bytes"]["max_claim_text_bytes"],
            max_sources_per_request=raw["counts"]["max_sources_per_request"],
            max_chunks_per_request=raw["counts"]["max_chunks_per_request"],
            max_citations=raw["counts"]["max_citations"],
            max_concurrent_model_calls=raw["counts"]["max_concurrent_model_calls"],
            max_chunk_overlap_lines=raw["counts"]["max_chunk_overlap_lines"],
            max_transient_retries=raw["counts"]["max_transient_retries"],
            max_claims_per_answer=raw["counts"]["max_claims_per_answer"],
            max_citation_ids_per_claim=raw["counts"]["max_citation_ids_per_claim"],
            max_format_retries=raw["counts"]["max_format_retries"],
            max_chunk_tokens=raw["tokens"]["max_chunk_tokens"],
            max_request_input_tokens=raw["tokens"]["max_request_input_tokens"],
            max_output_tokens_per_call=raw["tokens"]["max_output_tokens_per_call"],
            bytes_per_token_estimate=raw["tokens"]["bytes_per_token_estimate"],
            gate_probe_deadline_ms=raw["deadlines_ms"]["gate_probe"],
            spill_io_deadline_ms=raw["deadlines_ms"]["spill_io"],
            model_call_deadline_ms=raw["deadlines_ms"]["model_call"],
            request_deadline_ms=raw["deadlines_ms"]["request"],
            spill_ttl_seconds=raw["spill"]["ttl_seconds"],
            json_max_depth=raw["json"]["max_depth"],
            json_max_nodes=raw["json"]["max_nodes"],
            inspect_max_result_bytes=raw["inspect"]["max_result_bytes"],
            inspect_max_segments=raw["inspect"]["max_segments"],
            inspect_max_lines_per_page=raw["inspect"]["max_lines_per_page"],
            inspect_max_bytes_per_page=raw["inspect"]["max_bytes_per_page"],
            inspect_max_scan_lines=raw["inspect"]["max_scan_lines"],
            inspect_max_scan_bytes=raw["inspect"]["max_scan_bytes"],
            inspect_max_search_matches=raw["inspect"]["max_search_matches"],
            inspect_max_needle_bytes=raw["inspect"]["max_needle_bytes"],
            disclosure_max_per_source_bytes=raw["disclosure"]["max_per_source_bytes"],
            disclosure_max_per_session_bytes=raw["disclosure"]["max_per_session_bytes"],
            store_ddl_version=raw["store"]["ddl_version"],
            store_busy_timeout_ms=raw["store"]["busy_timeout_ms"],
            store_max_entries=raw["store"]["max_entries"],
            store_max_bytes=raw["store"]["max_bytes"],
            store_handle_ttl_seconds=raw["store"]["handle_ttl_seconds"],
            stats_max_records_per_page=raw["accounting"]["max_stats_records_per_page"],
            stats_max_pages=raw["accounting"]["max_stats_pages"],
        )

    def narrow(self, **overrides: int) -> Limits:
        """Return a copy with lowered caps. Raising a cap is a configuration error."""
        from dataclasses import replace

        for key, value in overrides.items():
            if not hasattr(self, key):
                raise ValueError(f"unknown limit: {key}")
            current = getattr(self, key)
            if (
                not isinstance(current, int)
                or isinstance(value, bool)
                or not isinstance(value, int)
            ):
                raise ValueError(f"limit is not numeric: {key}")
            if value > current:
                raise ValueError(f"limit {key} may only be narrowed (max {current})")
            if value < 0:
                raise ValueError(f"limit {key} may not be negative")
            if value == 0 and key in POSITIVE_LIMITS:
                raise ValueError(f"limit {key} must be positive")
        return replace(self, **overrides)


#: Caps a deployment may not set to zero, because zero would disable rather than tighten.
POSITIVE_LIMITS = frozenset(
    {
        "bytes_per_token_estimate",
        "max_chunk_bytes",
        "max_chunk_tokens",
        "max_concurrent_model_calls",
        "inspect_max_result_bytes",
        "inspect_max_scan_lines",
        "store_ddl_version",
        "store_busy_timeout_ms",
        "stats_max_records_per_page",
        "stats_max_pages",
    }
)

DEFAULT_LIMITS = Limits.defaults()
READER_MODEL = DEFAULT_LIMITS.reader_model
#: Deterministic estimator name recorded whenever provider usage is not available.
BASELINE_ESTIMATE_METHOD: str = raw_limits()["accounting"]["baseline_method"]


def legal_pair(status: str, code: str) -> bool:
    pairs = status_code_pairs()["pairs"]
    return status in pairs and code in pairs[status]


def supported_request_version(version: Any) -> bool:
    return isinstance(version, str) and version in SUPPORTED_REQUEST_VERSIONS


def envelope_byte_cap(result_kind: str | None, limits: Limits = DEFAULT_LIMITS) -> int:
    """The serialized cap that applies to one envelope.

    Deterministic extraction and stats carry a bounded payload of their own - up to
    ``max_extraction_bytes`` of exact snapshot bytes, or one page of operation records -
    so they are measured against ``max_extended_envelope_bytes``. Every other envelope
    keeps the original 16 KiB cap. Both values live in ``contracts/v1/limits.json``.
    """
    if result_kind in ("deterministic_extraction", "stats"):
        return limits.max_extended_envelope_bytes
    return limits.max_envelope_bytes
